fcrepo-import/src/main/java/org/fcrepo/importer/Importer.java

```diff
@@ -17,26 +17,178 @@
+            }
+        }
+    }
+
+    private void importFile(final File f) {
+        FcrepoResponse response = null;
+        URI uri = null;
+        try {
+            final Model model = createDefaultModel().read(new FileInputStream(f), null,
+                    config.getRdfLanguage());
+            final ResIterator binaryResources = model.listResourcesWithProperty(RDF_TYPE, NON_RDF_SOURCE);
+            if (binaryResources.hasNext()) {
+                uri = new URI(binaryResources.nextResource().getURI());
+                logger.info("Importing binary {}", f.getAbsolutePath());
+                response = importBinary(uri, sanitize(model));
+            } else {
+                uri = uriForFile(f, config.getDescriptionDirectory());
+                logger.info("Importing container {}", f.getAbsolutePath());
+                response = importContainer(uri, sanitize(model));
+            }
+
+            if (response.getStatusCode() > 204) {
+                logger.warn("Error while importing {} ({}): {}",
+                   f.getAbsolutePath(), response.getStatusCode(), IOUtils.toString(response.getBody()));
+            } else {
+                logger.info("Imported {}: {}", f.getAbsolutePath(), uri);
+            }
+        } catch (FcrepoOperationFailedException ex) {
+            logger.warn("Error importing {}: {}", f.getAbsolutePath(), ex.toString());
+            ex.printStackTrace();
+        } catch (IOException ex) {
+            logger.warn("Error reading or parsing {}: {}", f.getAbsolutePath(), ex.toString());
+            ex.printStackTrace();
+        } catch (URISyntaxException ex) {
+            logger.warn("Error building URI for {}: {}", f.getAbsolutePath(), ex.toString());
+            ex.printStackTrace();
+        } catch (Exception ex) {
+            ex.printStackTrace();
+        }
+    }
+    private FcrepoResponse importBinary(final URI binaryURI, final Model model)
+            throws FcrepoOperationFailedException, IOException {
+        final String contentType = model.getProperty(createResource(binaryURI.toString()), HAS_MIME_TYPE).getString();
+        final File binaryFile = fileForURI(binaryURI);
+        final FcrepoResponse binaryResponse = client.put(binaryURI).body(binaryFile, contentType).perform();
+        if (binaryResponse.getStatusCode() == 201) {
+            logger.info("Imported binary: {}", binaryURI);
+        } else {
+            return binaryResponse;
+        }
+
+        final URI descriptionURI = binaryResponse.getLinkHeaders("describedby").get(0);
+        return client.put(descriptionURI).body(modelToStream(model), config.getRdfLanguage())
+            .preferLenient().perform();
+    }
+
+    private FcrepoResponse importContainer(final URI uri, final Model model) throws FcrepoOperationFailedException {
+        return client.put(uri).body(modelToStream(model), config.getRdfLanguage()).preferLenient().perform();
+    }
+
+    private Model sanitize(final Model model) {
```

Comment:
This would be a natural for using lazy stream processing with `filters`... if you are interested.